pkg/cert-watcher/watcher.go

```diff
@@ -0,0 +1,155 @@
+		SecretName:      "k8tz-tls",
+		SecretNamespace: "k8tz",
+		Verbose:         false,
+	}
+}
+
+func (w *Watcher) Start(kubeconfigFlag string) error {
+	infoLogger.Println(version.DisplayVersion())
+
+	if w.Verbose {
+		verboseLogger.SetOutput(os.Stderr)
+		verboseLogger.Printf("server=%+v", *w)
+	}
+
+	infoLogger.Printf("Watching kubenetes secret: %s/%s", w.SecretNamespace, w.SecretName)
+	infoLogger.Printf("Syncing tls.crt on %s", w.TLSCertFile)
+	infoLogger.Printf("Syncing tls.key on %s", w.TLSKeyFile)
+
+	err := w.InitializeClientset(kubeconfigFlag)
+	if err != nil {
+		errorLogger.Printf("failed to setup connection with kubernetes api: %v", err)
+		return fmt.Errorf("failed to setup connection with kubernetes api: %w", err)
+	}
+
+	stopper := make(chan struct{})
+	defer close(stopper)
+
+	factory := informers.NewSharedInformerFactoryWithOptions(w.clientset, 0, informers.WithNamespace(w.SecretNamespace))
+	secretInformer := factory.Core().V1().Secrets().Informer()
+
+	defer runtime.HandleCrash()
+
+	go factory.Start(stopper)
+
+	if !cache.WaitForCacheSync(stopper, secretInformer.HasSynced) {
+		runtime.HandleError(fmt.Errorf("timed out waiting for secretInformer caches to sync"))
+		return fmt.Errorf("timed out waiting for secretInformer caches to sync")
+	}
+
+	secretInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			infoLogger.Printf("receiving add event on secret %s/%s", w.SecretNamespace, w.SecretName)
+			w.ProcessSecret(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			infoLogger.Printf("receiving update event on secret %s/%s", w.SecretNamespace, w.SecretName)
+			w.ProcessSecret(newObj)
+		},
+	})
+
+	<-stopper
+
+	return nil
+}
+
+func getKubeconfig(kubeconfPath string) (*restclient.Config, error) {
+	if kubeconfPath == "" {
+		verboseLogger.Println("--kubeconfig not specified. Using the inClusterConfig. This might not work.")
+		kubeconfig, err := restclient.InClusterConfig()
+		if err == nil {
+			return kubeconfig, nil
+		}
+
+		warningLogger.Println("error creating inClusterConfig, falling back to default config.")
+	}
+	return clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&clientcmd.ClientConfigLoadingRules{ExplicitPath: kubeconfPath},
+		&clientcmd.ConfigOverrides{ClusterInfo: clientcmdapi.Cluster{Server: ""}}).ClientConfig()
+}
+
+func (w *Watcher) InitializeClientset(kubeconfPath string) error {
+	config, err := getKubeconfig(kubeconfPath)
+	if err != nil {
+		return fmt.Errorf("failed to get in-cluster config: %v", err)
+	}
+
+	clientset, _ := kubernetes.NewForConfig(config)
+	if err != nil {
+		return fmt.Errorf("failed to create k8s client: %v", err)
+	}
+
+	w.clientset = clientset
+	return nil
+}
+
+func overwriteFile(filepath string, filecontent []byte) {
+	infoLogger.Printf("overwriting file %s", filepath)
+
+	fileCrt, err := os.Create(filepath)
+	if err != nil {
+		errorLogger.Printf("error creating file: %s, error=%v", filepath, err)
```

Comment:
its not enough just to print the error. it can lead to cert-key mismatch if we wrote the cert but failed to write the key for some reason (same for line 136).
i think the best we can do on such errors is to crash and make the pod restart. wdyt?

Reply:
instead of using Printf, can we use Panicf to crash the process and make the pod restart?